src/NoteDisplay.cpp

```diff
@@ -495,42 +786,165 @@ void NoteDisplay::DrawHoldPart( vector<Sprite*> &vpSpr, int iCol, int fYStep, fl
+				break;
+			case NCSM_Position:
+				column_args.zoom_handler->EvalForBeat(column_args.song_beat, cur_beat, sp_zoom);
+				render_width= fFrameWidth * sp_zoom[0];
+				break;
+		}
+
 		const float fFrameWidthScale	= ArrowEffects::GetFrameWidthScale( m_pPlayerState, fYOffset, fOverlappedTime );
-		const float fScaledFrameWidth	= fFrameWidth * fFrameWidthScale;
+		const float fScaledFrameWidth	= render_width * fFrameWidthScale;
 
-		float fX			= ArrowEffects::GetXPos( m_pPlayerState, iCol, fYOffset );
+		// Can't use the same code as for taps because hold bodies can only rotate
+		// around the y axis. -Kyz
+		switch(column_args.rot_handler->m_spline_mode)
+		{
+			case NCSM_Disabled:
+				// XXX: Actor rotations use degrees, Math uses radians. Convert here.
+				ae_rot[1]= ArrowEffects::GetRotationY(m_pPlayerState, fYOffset) * PI_180;
+				sp_rot.resize(3);
+				sp_rot[0]= sp_rot[1]= sp_rot[2]= 0.0f;
+				break;
+			case NCSM_Offset:
+				ae_rot[1]= ArrowEffects::GetRotationY(m_pPlayerState, fYOffset) * PI_180;
+				column_args.rot_handler->EvalForBeat(column_args.song_beat, cur_beat, sp_rot);
+				break;
+			case NCSM_Position:
+				column_args.rot_handler->EvalForBeat(column_args.song_beat, cur_beat, sp_rot);
+				break;
+		}
 
-		// XXX: Actor rotations use degrees, RageFastCos/Sin use radians. Convert here.
-		const float fRotationY		= ArrowEffects::GetRotationY( m_pPlayerState, fYOffset ) * PI/180;
+		RageVector3 center_vert(sp_pos[0] + ae_pos[0],
+			sp_pos[1] + ae_pos[1], sp_pos[2] + ae_pos[2]);
 
-		// if we're rotating, we need to modify the X and Z coords for the outer edges.
-		const float fRotOffsetX		= (fScaledFrameWidth/2) * RageFastCos(fRotationY);
-		const float fRotOffsetZ		= (fScaledFrameWidth/2) * RageFastSin(fRotationY);
+		// Special case for hold caps, which have the same top and bottom beat.
+		if(top_beat == bottom_beat && fY != fYStartPos)
+		{
+			center_vert.x+= render_forward.x;
+			center_vert.y+= render_forward.y;
+			center_vert.z+= render_forward.z;
+		}
 
-		//const float fXLeft			= fX - (fScaledFrameWidth/2);
-		const float fXLeft			= fX - fRotOffsetX;
-		const float fXCenter		= fX;
-		//const float fXRight		= fX + (fScaledFrameWidth/2);
-		const float fXRight		= fX + fRotOffsetX;
-		const float fZLeft			= fZ - fRotOffsetZ;
-		const float fZCenter	= fZ;
-		const float fZRight		= fZ + fRotOffsetZ;
+		const float render_roty= (sp_rot[1] + ae_rot[1]);
+
+		// (step 2 of vector handling)
+		RageVector3 render_left;
+		if(abs(render_forward.z) > 0.9f) // 0.9 arbitrariliy picked.
+		{
+			RageVec3Cross(&render_left, &render_forward, &pos_y_vec);
+		}
+		else
+		{
+			RageVec3Cross(&render_left, &render_forward, &pos_z_vec);
+		}
+		RageAARotate(&render_left, &render_forward, render_roty);
+		const float half_width= fScaledFrameWidth * .5f;
+		render_left.x*= half_width;
+		render_left.y*= half_width;
+		render_left.z*= half_width;
+
+		RageVector3 const left_vert(center_vert.x + render_left.x,
```

Comment:
Whoops, sometimes I mess up the order.  I prefer const to the right of the thing it's affecting because that's the way it binds when it's not the first token of the declaration.  "const foo const*" is nonsense, "foo const * const" is a const pointer to a const object.